modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java

```diff
@@ -648,4 +655,89 @@ private void ackAsciiLogo() {
     public static String cdcInstanceName(String igniteInstanceName) {
         return "cdc-" + igniteInstanceName;
     }
+
+    /** Iterator over {@link DataEntry}. */
+    public static class DataEntryIterator implements Iterator<DataEntry>, AutoCloseable {
+        /** WAL iterator. */
+        private final WALIterator walIter;
+
+        /** Last {@link DataRecord}.*/
+        private IgniteBiTuple<WALPointer, WALRecord> lastWalRec;
+
+        /** Iterator over {@link DataEntry}. */
+        private Iterator<DataEntry> entries;
+
+        /** Last {@link DataEntry} index inside last {@link DataRecord}. */
+        private int entryIdx;
+
+        /** @param walIter WAL iterator. */
+        DataEntryIterator(WALIterator walIter) {
+            this.walIter = walIter;
+        }
+
+        /** @return Current state. */
+        public CdcState state() {
+            return new CdcState(lastWalRec.get1(), entryIdx);
+        }
+
+        /** Initialize state. */
+        public void init(CdcState state) {
+            for (int i = 0; i < state.entryIndex(); i++) {
+                if (!hasNext())
+                    throw new IgniteException("Failed to restore entry index [rec=" + lastWalRec + ']');
+
+                next();
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean hasNext() {
+            advance();
+
+            return hasCurrent();
+        }
+
+        /** {@inheritDoc} */
+        @Override public DataEntry next() {
+            advance();
+
+            if (!hasCurrent())
+                throw new NoSuchElementException();
+
+            entryIdx++;
+
+            return entries.next();
+        }
+
+        /** */
+        private void advance() {
+            if (hasCurrent())
+                return;
+
+            while (walIter.hasNext()) {
+                lastWalRec = walIter.next();
+
+                DataRecord rec = (DataRecord)lastWalRec.get2();
+
+                entries = rec.writeEntries().iterator();
```

Comment:
This will create additional singletonList on each `DataRecord` that contains only one `DataEntry`.
Let's use `DataRecord#entryCount` and `DataRecord#get(int)` to iterate over data records.

Reply:
Fixed. Thanks.